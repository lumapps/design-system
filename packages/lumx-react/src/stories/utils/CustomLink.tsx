/* eslint-disable jsx-a11y/anchor-has-content,jsx-a11y/anchor-is-valid */
import React from 'react';

/**
 * Example custom link that can be used in linkAs props.
 */
export const CustomLink: any = ({ ...props }) => <a href="#" data-custom-link {...props} />;
