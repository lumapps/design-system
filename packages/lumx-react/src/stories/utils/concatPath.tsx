export type PathPart = string | number;
export type Path = PathPart | Array<PathPart>;

/**
 * Concat flatten object path
 *
 * @example concatPath('foo', 'bar') // => 'foo.bar'
 * @example concatPath(['foo', 0]) // => 'foo[0]'
 * @example concatPath('foo', 0, ['bar']) // => 'foo[0].bar'
 */
export const concatPath = (...prefix: Path[]) => {
    const [first, ...rest] = prefix.flat();
    return rest.reduce<string>((acc, part) => {
        if (typeof part === 'number') return `${acc}[${part}]`;
        return `${acc}.${part}`;
    }, String(first));
};
