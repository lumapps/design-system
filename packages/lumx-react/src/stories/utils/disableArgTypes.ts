export function disableArgTypes(argTypes: string[]) {
    return Object.fromEntries(argTypes.map((argType) => [argType, { table: { disable: true } }]));
}
