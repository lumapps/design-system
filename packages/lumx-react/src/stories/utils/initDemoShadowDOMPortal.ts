/** Init a demo shadow DOM to use as portal container */
export const initDemoShadowDOMPortal = () => {
    const div = document.createElement('div');
    document.body.appendChild(div);
    const container = div.attachShadow({ mode: 'open' });
    const style = document.createElement('style');
    style.innerText = `button { color: red; }`;
    container.appendChild(style);
    return { container, teardown: () => div.remove() };
};
