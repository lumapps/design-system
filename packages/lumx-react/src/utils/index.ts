/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
