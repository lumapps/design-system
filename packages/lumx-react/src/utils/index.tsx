export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
