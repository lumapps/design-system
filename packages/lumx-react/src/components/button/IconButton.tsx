import React from 'react';

import { Emphasis, Icon, Size, Theme, Tooltip } from '@lumx/react';
import { BaseButtonProps, ButtonRoot } from '@lumx/react/components/button/ButtonRoot';
import { COMPONENT_PREFIX } from '@lumx/react/constants';
import { Comp, getRootClassName } from '@lumx/react/utils';

/**
 * Defines the props of the component.
 */
export interface IconButtonProps extends BaseButtonProps {
    /**
     * The icon name to use as the button label.
     * @see {@link IconProps#icon}
     */
    icon: string;
    /**
     * The label of the tooltip. It is required for a11y purpose.
     * If you really don't want a tooltip and aria-label, you can give an empty label (this is not recommended).
     */
    label: string;
}

/**
 * The display name of the component.
 */
const COMPONENT_NAME = `${COMPONENT_PREFIX}IconButton`;

/**
 * The default class name and classes prefix for this component.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * The default value of props.
 */
const DEFAULT_PROPS: Partial<IconButtonProps> = {
    emphasis: Emphasis.high,
    size: Size.m,
    theme: Theme.light,
};

export const IconButton: Comp<IconButtonProps> = (props) => {
    const { emphasis, icon, label, size, theme, ...forwardedProps } = props;

    return (
        <Tooltip label={label}>
            <ButtonRoot {...{ emphasis, size, theme, ...forwardedProps }} aria-label={label} variant="icon">
                <Icon icon={icon} />
            </ButtonRoot>
        </Tooltip>
    );
};
IconButton.displayName = COMPONENT_NAME;
IconButton.className = CLASSNAME;
IconButton.defaultProps = DEFAULT_PROPS;
