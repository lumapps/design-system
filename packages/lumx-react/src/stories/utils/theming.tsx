import React from 'react';
import { Predicate } from '@lumx/react/utils/type';
import { FlexBox } from '@lumx/react';

type Property = string | { value: string };
interface PropertyTree {
    [key: string]: PropertyTree | Property;
}
type Entry = [key: string, value: string];
type Variables = Record<string, string>;

function asValue(prop?: PropertyTree | Property): string | undefined {
    if (typeof prop === 'string') {
        return prop;
    }
    if (typeof prop === 'object' && 'value' in prop) {
        return prop.value as string;
    }
    return undefined;
}

/** Convert property tree into a list of args entries ([key, value]) */
function* toArgEntries(props: PropertyTree | Property, parent = ''): Iterable<Entry> {
    const value = asValue(props);
    if (value) {
        yield [parent, value];
    } else {
        for (const [key, prop] of Object.entries(props)) {
            const newParent = parent ? `${parent}-${key}` : key;
            yield* toArgEntries(prop, newParent);
        }
    }
}
function* parseVariableEntries(cssString: string, strict = false) {
    for (const line of cssString.split(/(?<=;)/)) {
        if (!line.trim()) continue;
        const match = line.match(/^\s*(--lumx[^:]+)\s*:\s*([^;]+);/m);
        if (!match) {
            if (strict) throw new Error(`Could not parse \`${line}\``);
            else continue;
        }
        const [, variableName, variableValue] = match;
        yield [variableName, variableValue];
    }
}

function VariableEditor({
    variable: [key, value],
    setVariable,
    defaultStyle,
}: {
    variable: Entry;
    setVariable: (newVariable: Entry) => void;
    defaultStyle: Variables;
}) {
    const onChange = React.useCallback<React.ChangeEventHandler<HTMLInputElement>>(
        (evt) => {
            setVariable([evt.target.name, evt.target.value]);
        },
        [setVariable],
    );

    return (
        <FlexBox as="label" orientation="horizontal" hAlign="center" style={{ whiteSpace: 'nowrap' }}>
            <span style={{ flexShrink: 0 }}>{key}</span>:
            <input
                name={key}
                onChange={onChange}
                style={{ border: 0, padding: 8, fieldSizing: 'content', display: 'inline' } as any}
                value={defaultStyle[key] === value ? '' : value}
                placeholder={value}
            />
            ;
        </FlexBox>
    );
}

function VariablesEditor({
    style,
    defaultStyle,
    setStyle,
}: {
    style: Variables;
    defaultStyle: Variables;
    setStyle: (fn: (newStyle: Variables) => Variables) => void;
}) {
    const setVariable = React.useCallback(
        ([key, value]: Entry) => {
            setStyle((prevStyle) => ({ ...prevStyle, [key]: value }));
        },
        [setStyle],
    );
    return (
        <details style={{ fontFamily: 'monospace' }} className="lumx-spacing-margin-vertical-huge">
            <summary className="lumx-typography-subtitle2">Edit variables</summary>
            {Object.entries(style).map((variable) => (
                <VariableEditor
                    key={variable[0]}
                    defaultStyle={defaultStyle}
                    variable={variable}
                    setVariable={setVariable}
                />
            ))}
        </details>
    );
}

const FILTER_PROPS = ['theme', 'emphasis', 'size'];

export function withTheming<P extends PropertyTree>({
    properties,
    values,
}: {
    properties?: P;
    values?: string | Partial<P>;
}) {
    const initialStyle = values
        ? Object.fromEntries(
              typeof values === 'string'
                  ? parseVariableEntries(values)
                  : toArgEntries(values as PropertyTree, '--lumx'),
          )
        : [];
    const defaultStyle = Object.fromEntries(properties ? Array.from(toArgEntries(properties, '--lumx')) : []);

    const varNames = new Set(Object.keys(initialStyle).concat(Object.keys(defaultStyle)));

    function filteredStyles(args: any) {
        // Create predicate
        const filterArgs = Object.entries(args)
            .map(([argName, argValue]) => {
                if (typeof argValue === 'string' && FILTER_PROPS.includes(argName)) {
                    return (varName: string) =>
                        // Has the given arg with value
                        varName.match(new RegExp(`${argName}-${argValue}`)) ||
                        (argName === 'emphasis' && argValue === 'medium' && varName.match(/emphasis-selected/)) ||
                        // Or does not have the arg at all
                        !varName.match(new RegExp(`${argName}-\\w+`));
                }
                return false;
            })
            .filter(Boolean) as Array<Predicate<string>>;
        const styles = Object.fromEntries(
            Array.from(varNames.values())
                .map((varName) => {
                    if (!filterArgs.every((filter) => filter(varName))) return null as any;
                    return [varName, initialStyle[varName] || defaultStyle[varName]];
                })
                .filter(Boolean),
        );
        return () => styles;
    }

    return (Story: any, ctx: any) => {
        const [style, setStyle] = React.useState<Variables>(filteredStyles(ctx.args));

        return (
            <>
                <Story args={{ ...ctx.args, style: { ...ctx.args.style, ...style } }} />
                <VariablesEditor style={style} defaultStyle={defaultStyle} setStyle={setStyle} />
            </>
        );
    };
}
