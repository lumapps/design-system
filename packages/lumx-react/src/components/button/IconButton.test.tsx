import React, { ReactElement } from 'react';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';

import { commonTestsSuite, Wrapper } from '@lumx/react/testing/utils';

import { IconButton, IconButtonProps } from './IconButton';

const DEFAULT_PROPS = IconButton.defaultProps as any;
const CLASSNAME = IconButton.className as string;

type SetupProps = Partial<IconButtonProps>;

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = (propsOverride: SetupProps = {}, shallowRendering = true) => {
    const props: any = { ...propsOverride };
    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;

    const wrapper: Wrapper = renderer(<IconButton {...props} />);

    return {
        buttonRoot: wrapper.find('ButtonRoot'),
        icon: wrapper.find('Icon'),
        props,
        wrapper,
    };
};

describe(`<${IconButton.displayName}>`, () => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', () => {
        it('should render icon button', () => {
            const { buttonRoot, icon, wrapper } = setup({});
            expect(wrapper).toMatchSnapshot();
            expect(buttonRoot).toExist();
            expect(icon).toExist();
        });
    });

    // 2. Test defaultProps value and important props custom values.
    describe('Props', () => {
        it('should use default props', () => {
            const { wrapper, buttonRoot } = setup();
            expect(wrapper).toMatchSnapshot();

            const actualProps = buttonRoot.props() as IconButtonProps;
            expect(actualProps.variant).toEqual('icon');
            for (const [propName, propValue] of Object.entries(DEFAULT_PROPS)) {
                expect(actualProps[propName]).toEqual(propValue);
            }
        });

        it('should forward any CSS class', () => {
            const props: Partial<IconButtonProps> = {
                className: 'component component--is-tested',
            };
            const { wrapper, buttonRoot } = setup(props);
            expect(wrapper).toMatchSnapshot();

            expect(buttonRoot).toHaveClassName(props.className);
        });
    });

    // 3. Test events.
    describe('Events', () => {
        // Nothing to do here.
    });

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', () => {
        // Nothing to do here.
    });

    // 5. Test state.
    describe('State', () => {
        // Nothing to do here.
    });

    // Common tests suite.
    commonTestsSuite(setup, { prop: 'buttonRoot' }, { className: CLASSNAME });
});
