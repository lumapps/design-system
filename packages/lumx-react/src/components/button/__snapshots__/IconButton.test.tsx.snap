// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconButton> Props should forward any CSS class 1`] = `
<Tooltip
  delay={500}
  placement="bottom"
>
  <ButtonRoot
    className="component component--is-tested"
    emphasis="high"
    size="m"
    theme="light"
    variant="icon"
  >
    <Icon />
  </ButtonRoot>
</Tooltip>
`;

exports[`<IconButton> Props should use default props 1`] = `
<Tooltip
  delay={500}
  placement="bottom"
>
  <ButtonRoot
    emphasis="high"
    size="m"
    theme="light"
    variant="icon"
  >
    <Icon />
  </ButtonRoot>
</Tooltip>
`;

exports[`<IconButton> Snapshots and structure should render icon button 1`] = `
<Tooltip
  delay={500}
  placement="bottom"
>
  <ButtonRoot
    emphasis="high"
    size="m"
    theme="light"
    variant="icon"
  >
    <Icon />
  </ButtonRoot>
</Tooltip>
`;
